x-pack/solutions/security/plugins/security_solution/server/lib/telemetry/diagnostic/health_diagnostic_receiver.ts

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { filter, finalize, merge, type Observable, Subject, takeUntil, tap, timer } from 'rxjs';
+import * as rx from 'rxjs';
+import type { ElasticsearchClient, LogMeta, Logger } from '@kbn/core/server';
+import type { EqlSearchRequest, SearchRequest } from '@elastic/elasticsearch/lib/api/types';
+import type { QueryConfig, CircuitBreakingQueryExecutor } from './health_diagnostic_receiver.types';
+import type {
+  CircuitBreaker,
+  CircuitBreakerResult,
+} from './health_diagnostic_circuit_breakers.types';
+import { type HealthDiagnosticQuery, QueryType } from './health_diagnostic_service.types';
+import type { TelemetryLogger } from '../telemetry_logger';
+import { newTelemetryLogger } from '../helpers';
+
+export class CircuitBreakingQueryExecutorImpl implements CircuitBreakingQueryExecutor {
+  private readonly logger: TelemetryLogger;
+
+  constructor(private client: ElasticsearchClient, logger: Logger) {
+    this.logger = newTelemetryLogger(logger.get('circuit-breaking-query-executor'));
+  }
+
+  async search<T>(queryConfig: QueryConfig): Promise<T[]> {
+    const { query, circuitBreakers } = queryConfig;
+
+    const upstream$ = new Subject<T[]>();
+    const stop$ = new Subject<void>();
+
+    switch (query.type) {
+      case QueryType.DSL:
+        this.streamDsl(query, upstream$, stop$).catch((err: unknown) => {
+          upstream$.error(err);
+        });
+        break;
+      case QueryType.EQL:
+        this.streamEql(query, upstream$, stop$).catch((err: unknown) => {
+          upstream$.error(err);
+        });
+        break;
+      case QueryType.ESQL:
+        this.streamEsql(query, upstream$, stop$).catch((err: unknown) => {
+          upstream$.error(err);
+        });
+        break;
+      default: {
+        const exhaustiveCheck: never = query.type;
+        throw new Error(`Unhandled QueryType: ${exhaustiveCheck}`);
+      }
+    }
+
+    return rx.firstValueFrom(
+      upstream$.pipe(
+        takeUntil(this.configureCircuitBreakers(circuitBreakers)),
+        finalize(() => {
+          stop$.next();
+        })
+      )
+    );
+  }
+
+  async streamDsl<T>(
+    diagnosticQuery: HealthDiagnosticQuery,
+    collector$: Subject<T[]>,
+    stop$: Subject<void>
+  ): Promise<void> {
+    const controller = new AbortController();
+    const abortSignal = controller.signal;
+    const index = await this.indexFor(diagnosticQuery);
+
+    if (index.length === 0) {
+      this.logger.warn('No indices found for the query', {
+        queryName: diagnosticQuery.name,
+      } as LogMeta);
+      collector$.next([]);
+      return;
+    }
+
+    const stopSub: rx.Subscription = stop$.subscribe(() => {
+      controller.abort();
+    });
```

Comment:
nit: IIUC this block duplicates itself in all three functions. We could move it one level up (to the `search` function, including try-finally) and leave only logic inside try in the each function. That will also remove dependency of these methods on `stop$`.